src/tools/transactions_builder.rs

```diff
@@ -0,0 +1,255 @@
+    }
+
+    pub fn stake(
+        &mut self,
+        token: ContractName,
+        staking_contract: ContractName,
+        amount: u128,
+    ) -> Result<(), Error> {
+        self.add_stake_cf(StakingAction::Stake { amount });
+        self.add_hyllar_cf(
+            token,
+            ERC20Action::Transfer {
+                recipient: staking_contract.0.clone(),
+                amount,
+            },
+            None,
+        );
+
+        Ok(())
+    }
+
+    pub fn delegate(&mut self, validator: ValidatorPublicKey) -> Result<(), Error> {
+        self.add_stake_cf(StakingAction::Delegate { validator });
+
+        Ok(())
+    }
+
+    pub async fn build(&mut self, states: &mut States) -> Result<BuildResult> {
+        let mut new_states = states.clone();
+        let mut outputs = vec![];
+        for id in self.hydentity_cf.iter() {
+            let runner = ContractRunner::new(
+                "hydentity".into(),
+                get_binary("hydentity".into())?,
+                self.identity.clone(),
+                id.1 .0.clone(),
+                self.blobs.clone(),
+                id.2.clone(),
+                new_states.hydentity.clone(),
+            )
+            .await?;
+            let out = runner.execute()?;
+            new_states.hydentity = out.next_state.clone().try_into()?;
+            outputs.push(("hydentity".into(), out));
+            self.runners.push(runner);
+        }
+
+        for cf in self.hyllar_cf.iter() {
+            let runner = ContractRunner::new(
+                cf.1.clone(),
+                get_binary(cf.1.clone())?,
+                self.identity.clone(),
+                BlobData(vec![]),
+                self.blobs.clone(),
+                cf.2.clone(),
+                new_states.for_token(&cf.1)?.clone(),
+            )
+            .await?;
+            let out = runner.execute()?;
+            new_states.update_for_token(&cf.1, out.next_state.clone().try_into()?)?;
+            outputs.push(("hyllar".into(), out));
+            self.runners.push(runner);
+        }
+
+        for cf in self.staking_cf.iter() {
+            info!("State before runner: {:?}", new_states.staking);
+            info!("on chain state: {:?}", new_states.staking.on_chain_state());
+            let runner = ContractRunner::new::<staking::state::OnChainState>(
+                "staking".into(),
+                get_binary("staking".into())?,
+                self.identity.clone(),
+                BlobData(new_states.staking.as_digest().0),
+                self.blobs.clone(),
+                cf.1.clone(),
+                new_states.staking.on_chain_state(),
+            )
+            .await?;
+            let out = runner.execute()?;
+            let mut contract = StakingContract::new(
+                ExecutionContext {
+                    callees_blobs: vec![].into(),
+                    caller: self.identity.clone(),
```

Comment:
OPT: je pense que c'est pas nécessaire de mettre un caller non?

Reply:
je comprends pas, tu voudrais que je mette quoi ?

Reply:
Rien !
Par défaut si on met rien on vérifie que pour l'identité `id.<nom_de_contract_d'identité>` il y a un blob pour le contract `<nom_de_contract_d'identité>`

Reply:
Ici, c'est une "copie" de ce qui ce passe dans le guest. C'est un petit tricks pour réussir à générer le même state que si on passait par le prover. Je passe pas par les blobs, j'execute le code du contrat en "natif", il n'y a pas de verification d'identité là, je cherche juste à re-générer le même state, car on-chain on a que le hash. J'ai besoin de build le vrai "next_state" pour pouvoir itérer sur mon builder....

FYI: ExecutionContext est défini comme ça:
```rust
pub struct ExecutionContext {
    pub callees_blobs: RefCell<Vec<Blob>>,
    pub caller: Identity,
}
```
j'ai pas le choix de passer un caller, et c'est tout à fait normal :)

Reply:
pas certain de la viabilité du truc, mais on peut pas passer par les indexers ?

Reply:
>mais on peut pas passer par les indexers ?

Non, ici on est pré-genesis quand on s'en sert, donc les indexers ne nous sont d'aucune utilité

Reply:
le builder il sert à construire une transaction pouvant contenir autant de blobs qu'on veut dedans (pour le même contract_name), il permet de build les blobs+proofs de manière cohérente en connaissant chaque state intermédiaire au sein de la blobTx.

Reply:
En soit la logique de "process le state d'une TX pour obtenir le state intermédiaire" c'est la logique des indexer, ce que je me demandais c'est s'il y avait pas un peu d'overlap fonctionnel entre les deux. Mais ça mérite clairement plus de réflexion.

Reply:
oui clairement il y a un overlap entre le TransactionBuilder et le ContractStateIndexer oui, j'aimerais réussir à rassembler les deux un jour